lib/github_service/commands/run_test.rb

```diff
@@ -0,0 +1,302 @@
+      end
+
+      def normalize_repo_name(repo)
+        repo.include?("/") ? repo : "#{issue.organization_name}/#{repo}"
+      end
+
+      def valid?
+        unless issue.pull_request?
+          issue.add_comment("@#{issuer} 'run-test(s)' command is only valid on pull requests, ignoring...")
+          return false
+        end
+
+        unless repos_valid?(@test_repos + @repos)
+          issue.add_comment("@#{issuer} '#{value}' is an invalid repo, ignoring...")
+
+          return false
+        end
+
+        true
+      end
+
+      REPO_ONLY_REGEXP = /(?:[^\/]+\/)?(?<REPO_NAME_ONLY>[^@#]+)/.freeze
+      def repos_valid?(repos)
+        repos.all? do |repo|
+          name_only = repo.match(REPO_ONLY_REGEXP)[:REPO_NAME_ONLY]
+          GithubService.org_repos(issue.organization_name).include?(name_only)
+        end
+      end
+
+      ##### run_tests steps #####
+
+      # Clone repo (if needed) and initialize @rugged_repo
+      def clone_repo_if_necessary
+        repo_path = self.class.test_repo_clone_dir
+        if Dir.exist?(self.class.test_repo_clone_dir)
+          @rugged_repo = Rugged::Repository.new(repo_path)
+        else
+          url = self.class.test_repo_url
+          @rugged_repo = Rugged::Repository.clone_at(url, repo_path)
+        end
+        git_fetch
+      end
+
+      def create_cross_repo_test_branch
+        rugged_repo.create_branch(branch_name, "origin/master")
+        rugged_repo.checkout(branch_name)
+      end
+
+      def update_travis_yaml_content
+        Dir.chdir(@rugged_repo.workdir) do
+          content = YAML.load_file(".travis.yml")
+
+          content["env"] = {} unless content["env"]
+          content["env"]["global"] = ["REPOS=#{repos.join(',')}"]
+          content["env"]["matrix"] = test_repos.map { |repo| "TEST_REPO=#{repo}" }
+
+          File.write('.travis.yml', content.to_yaml)
+        end
+      end
+
+      def commit_travis_yaml_changes
+        index = rugged_repo.index
+        index.add('.travis.yml')
+        index.write
+
+        # rubocop:disable: Rails/TimeZone
+        bot       = self.class.bot_name
+        author    = { :name => issuer, :email => "no-name@example.com", :time => Time.now }
+        committer = { :name => bot,    :email => "#{bot}@manageiq.org", :time => Time.now }
+        # rubocop:enable: Rails/TimeZone
+
+        Rugged::Commit.create(
+          rugged_repo,
+          :author     => author,
+          :committer  => committer,
+          :parents    => [rugged_repo.last_commit].compact,
+          :tree       => index.write_tree(rugged_repo),
+          :update_ref => "HEAD",
+          :message    => <<~COMMIT_MSG
+            Running tests for #{issuer}
+
+            From Pull Request:  #{issue.fq_repo_name}##{issue.number}
+          COMMIT_MSG
+        )
+      end
+
+      def push_commit_to_remote
+        push_options = {}
+
+        if Settings.github_credentials.username && Settings.github_credentials.password
```

Comment:
What happens if we don't have creds?  Will this blow up?

Reply:
Maybe [valid?](https://github.com/ManageIQ/miq_bot/pull/454/files#diff-2e4738d8e7840e1f2ad4b4161e5ddf2cR182) should check for this instead?  Not sure the best place to check for creds.

Reply:
Nope, since the tests I am running don't have creds either :)

Basically, it is similar to the implementation of `Repo#git_fetch` you wrote:

https://github.com/ManageIQ/miq_bot/blob/bc86fdec523510fc9c98fe363c708139a936feda/app/models/repo.rb#L133-L144

And will only add the options if they exist.  This means for clones that don't have credentials associated with them (like the file path ones I am using in the specs), this works when no credentials exist.

Granted, it will blow up if we are not authenticated with `github` properly (my current issue I am working through with [my testing at the moment](https://github.com/GrumpyTech/manageiq-ui-classic/pull/1#issuecomment-562637555)).  Was trying to avoid setting up my bot with a separate SSH key, but it might be unavoidable, and we might need to go that route here instead, and assume a `ssh` based clone instead.

Reply:
As ok...makes sense :+1:

Reply:
> Granted, it will blow up if we are not authenticated with `github` properly (my current issue I am working through with [my testing at the moment](https://github.com/GrumpyTech/manageiq-ui-classic/pull/1#issuecomment-562637555)). 

Okay, as mentioned here:

https://github.com/ManageIQ/miq_bot/pull/454#issuecomment-562762968

This was user (configuration) error, and works as expected when the bot has permissions write to that repo directly.

To avoid extra credentials (at least for dev), I think I am going to say this is good for now and resolve.  That said, I am fine with push back if you really want SSH key support instead, and I can look into adding that for this PR.  Just figured it is kind of redundant when we are already using user/pass auth with `octokit`.